infra/discovery/src/listener/listener.js

```diff
@@ -101,66 +106,90 @@ const config = {
  */
 async function main() {
   const context = await new Context().init(config, errorCounter)
+
   let processedToBlock = await getLastBlock(context.config)
+  logger.info(`Resuming processing from ${processedToBlock}`)
 
-  logger.info(`Resuming processing from block ${processedToBlock}`)
+  if (context.config.concurrency > 1) {
+    logger.warn(`Backfill mode: concurrency=${context.config.concurrency}`)
+  }
 
-  const checkIntervalSeconds = 5
+  // Helper function to wait at most tickIntervalSeconds.
+  const tickIntervalSeconds = 5
   let start
+  async function nextTick() {
+    const elapsed = new Date() - start
+    const delay = Math.max(tickIntervalSeconds * 1000 - elapsed, 1)
+    return new Promise(resolve => setTimeout(resolve(true), delay))
+  }
 
-  const check = async () => {
-    await withRetrys(async () => {
-      start = new Date()
-      const currentBlock = await context.web3.eth.getBlockNumber()
-      // Respect the trailBlocks option
-      const toBlock = Math.max(currentBlock - context.config.trailBlocks, 0)
-      if (toBlock === processedToBlock) {
-        logger.debug('No new blocks to process')
-        return scheduleNextCheck()
-      }
-      logger.debug(`Processing to block ${toBlock}`)
-      contractsContext.marketplace.eventCache.updateBlock(toBlock)
-      contractsContext.identityEvents.eventCache.updateBlock(toBlock)
-
-      // Retrieve all events for the relevant contracts up to the the block
-      // specified in the previous calls to updateBlock
-      const eventArrays = await Promise.all([
-        contractsContext.marketplace.eventCache.allEvents(),
-        contractsContext.identityEvents.eventCache.allEvents()
-      ])
-
-      // Flatten array of arrays filtering out anything undefined
-      const events = [].concat(...eventArrays.filter(x => x))
-      // Filter to only new events
-      const newEvents = events.filter(
-        event => event.blockNumber > processedToBlock
-      )
-      logger.debug(`Got ${newEvents.length} new events`)
-
+  do {
+    start = new Date()
+
+    // Compute the range of blocks to process,
+    // while respecting trailing block configuration.
+    const currentBlock = await context.web3.eth.getBlockNumber()
+    let toBlock = Math.max(currentBlock - context.config.trailBlocks, 0)
+    toBlock = Math.min(toBlock, processedToBlock + 10000) // At most process a chunk of 10k blocks at once.
```

Comment:
The eventCache code out of `@origin/graphql` has its own chunking logic, so perhaps this is not needed here

Reply:
My intent was so that in backfill mode where large number of blocks need to be processed, we would persist the progress in the listener table regularly. But perhaps not worth the extra complexity. Removed.